src/Microsoft.AspNetCore.Blazor.Server/Circuits/RenderBatchFormatterResolver.cs

```diff
@@ -29,15 +34,54 @@ public RenderBatch Deserialize(byte[] bytes, int offset, IFormatterResolver form
 
             public int Serialize(ref byte[] bytes, int offset, RenderBatch value, IFormatterResolver formatterResolver)
             {
-                using (var memoryStream = new MemoryStream())
+                // WARNING: Be careful never to execute .GetBuffer() or .ToArray() on this memory
+                // stream we get from recyclableMemoryStreamManager. If you do that, you'll silently
+                // eliminate all the benefit of RecyclableMemoryStream.
+                using (var memoryStream = recyclableMemoryStreamManager.GetStream($"{nameof(RenderBatchFormatter)}.{nameof(Serialize)}"))
                 using (var renderBatchWriter = new RenderBatchWriter(memoryStream, leaveOpen: false))
                 {
                     renderBatchWriter.Write(value);
-                    
-                    var bytesBuffer = memoryStream.GetBuffer();
-                    return MessagePackBinary.WriteBytes(ref bytes, offset, bytesBuffer, 0, (int)memoryStream.Length);
+
+                    var compressedStream = CompressStream(memoryStream);
+                    try
+                    {
+                        return CopyStreamToMessagePack(compressedStream, ref bytes, offset);
+                    }
+                    finally
+                    {
+                        compressedStream.Dispose();
+                    }
+                }
+            }
+
+            private static Stream CompressStream(Stream input)
+            {
+                // WARNING: Be careful never to execute .GetBuffer() or .ToArray() on this memory
+                // stream we get from recyclableMemoryStreamManager. If you do that, you'll silently
+                // eliminate all the benefit of RecyclableMemoryStream.
+                var output = recyclableMemoryStreamManager.GetStream($"{nameof(RenderBatchFormatter)}.{nameof(CompressStream)}");
+
+                using (var gzipStream = new GZipStream(output, CompressionLevel.Optimal, leaveOpen: true))
+                {
+                    input.Seek(0, SeekOrigin.Begin);
+                    input.CopyTo(gzipStream);
+                    gzipStream.Flush();
+                    return output;
```

Comment:
Given that some folks have concerns about GZip, is it worth making it optional?

Reply:
For the CPU/bandwidth I agree there are legitimate reasons to go either way, but I’d like to see clearer measurements to prove the CPU cost could ever be prohibitive. We can always add more options later. It’s hard to see in what scenarios people would realistically turn this off. 

As for security, I’d like to get a really clear conclusion from our discussions with security people. As it stands I expect we will be able to completely eliminate the entire class of compression oracle attacks, as we have so much more control than in the more general HTTP response case.